File: Spark/Pyspark/DataFrame_native_API_solutions/SuperStoreUS/src/main/python/SaleProfit_US_states.py
import sys
import configparser as config
import datetime

try:
    from pyspark.sql import SparkSession
    from pyspark.sql.types import DoubleType
    from pyspark.sql.types import IntegerType
    from pyspark.sql.functions import *

    props = config.RawConfigParser()
    props.read("src/main/configs/app.properties")
    exMode = sys.argv[1]

    spark = SparkSession.\
        builder.\
        master(props.get(exMode, 'deployMode')).\
        appName("Sale Profit Per State in US").\
        getOrCreate()

    sc = spark.sparkContext

    # Handling with file system
    dataInPath = props.get(exMode, 'data.in.path')
    dataOutPath = props.get(exMode, 'data.out.path')

    # HDFS API's
    Path = sc._gateway.jvm.org.apache.hadoop.fs.Path
    FileSytem = sc._gateway.jvm.org.apache.hadoop.fs.FileSystem
    Configuration = sc._gateway.jvm.org.apache.hadoop.conf.Configuration

    fs = FileSytem.get(Configuration())

    if (fs.exists(Path(dataInPath)) == True):
        if (fs.exists(Path(dataOutPath))):
            dataOutPath = dataOutPath + "_" + str(datetime.datetime.now().strftime("%y_%m_%d_%H_%M"))
            print("New Path = ", dataOutPath)

        dataDF = spark.read.csv(dataInPath,
                            header= True, inferSchema= True)

        #Type casting - since the inferScheame option doesn't properly infer the data type
        dataDF = dataDF.withColumn("Sales", dataDF.Sales.cast(DoubleType())).\
            withColumn("Quantity", dataDF.Quantity.cast(IntegerType())).\
            withColumn("Discount", dataDF.Discount.cast(DoubleType()))

        dataAggDF = dataDF.select(dataDF.Country, dataDF.State, dataDF.Profit).\
            groupBy("Country", "State").\
            agg(round(sum(dataDF.Profit), 2).alias("Total_profit"))

        dataAggDF.orderBy(dataAggDF.Total_profit.desc()).\
            coalesce(2).\
            write.\
            csv(dataOutPath, mode= "overwrite")
    else:
        print("Input Path is not exist" + dataInPath)

except ImportError as ie:
    print("Import Error with Spark modules", ie)
    sys.exit(1)

File: Spark/Spark_with_scala/DataFrame_native_API_solutions/SuperStore/src/main/scala/SaleProfit_US_states.scala
import java.util.Calendar

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions.col
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._
import com.typesafe.config.{Config, ConfigFactory}
import org.apache.hadoop.fs._

object SaleProfit_US_states {
  def main(args: Array[String]): Unit = {
    val props =ConfigFactory.load()
    val exMode = args(0)

    val spark = SparkSession.
      builder().
      master(props.getConfig(exMode).getString("deployMode")).
      appName("Sale Profit Per State in US").
      getOrCreate()

    //Handling with file system
    val dataInPath = props.getConfig(exMode).getString("data.in.path")
    var dataOutPath = props.getConfig(exMode).getString("data.out.path")

    //HDFS API's

    val sc = spark.sparkContext
    val fs = FileSystem.get(sc.hadoopConfiguration)

    if(fs.exists(new Path(dataInPath)) == true){
      if(fs.exists(new Path(dataOutPath))){
        val hour = Calendar.getInstance().get(Calendar.HOUR_OF_DAY)
        val day = Calendar.getInstance().get(Calendar.DAY_OF_MONTH)
        val month = Calendar.getInstance().get(Calendar.MONTH)+1
        val year = Calendar.getInstance().get(Calendar.YEAR)
        dataOutPath = dataOutPath+"_"+year+"_"+month+"_"+day+"_"+hour
      }
      val dataDF = spark.read.format("csv").
        option("sep", ",").
        option("inferSchema", true).
        option("header", true).
        load(dataInPath)

      //  Type casting - since the inferScheame option doesn't properly infer the data type
      //  df.withColumn(<old column name>,df.col(<dummy column>).cast(<datatype>))
      val dataDF1 = dataDF.
        withColumn("Sales", dataDF.col("Sales").cast(DoubleType)).
        withColumn("Quantity", dataDF.col("Quantity").cast(IntegerType)).
        withColumn("Discount", dataDF.col("Discount").cast(DoubleType))

      val dataAggDF = dataDF1.
        select(col("Country"), col("State"), col("Profit")).
        groupBy(col("Country"), col("State")).
        agg(round(sum((col("Profit"))), 2).alias("Total_profit"))

      dataAggDF.orderBy(col("Total_profit").desc).
        coalesce(2).
        write.csv(dataOutPath)
    }
    else {
      println("Input Path is not exist" + dataInPath)
    }
  }

}
